src/sql/magic.py

```diff
@@ -619,6 +622,73 @@ def _persist_dataframe(
         display.message_success(f"Success! Persisted {table_name} to the database.")
 
 
+def find_toml_path():
+    """
+    Recursively finds an absolute path to pyproject.toml starting
+    from current directory
+    """
+    current = Path().resolve()
+    while not (current / "pyproject.toml").exists():
+        if current == current.parent:
+            return None
+
+        current = current.parent
+    display.message(f"Found pyproject.toml from '{current}'")
+
+    return str(current) + "/pyproject.toml"
+
+
+def get_configs(file_path):
+    """Returns saved configs from given file_path"""
+    with open(file_path, "r") as file:
+        content = file.read()
+        data = toml.loads(content)
+    try:
+        return data["tool"]["jupysql"]["SqlMagic"]
+    except KeyError:
+        return {}
+
+
+def get_config_variables(obj):
+    """Returns a list of configs users can set"""
+    config_vars = list(vars(obj)["_all_trait_default_generators"].keys())
```

Comment:
Should we infer the settings from the object or it should be declared in some file? what do you think @edublancas

Reply:
we should get them from the object so we don't have to maintain a separate list, but I don't think this is the best way to get them. in the `SqlMagic.__init__` method, we can grab the traits information

I haven't looked at this PR in detail but I'm assuming that this logic will be added in the `SqlMagic.__init__` which is executed when we execute `%load_ext sql`

```
In [1]: %load_ext sql
Python 3.10.12 (main, Jul  5 2023, 15:02:25) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: self.trait_defaults()
Out[1]:
{'autocommit': True,
 'autolimit': 0,
 'autopandas': False,
 'autopolars': False,
 'column_local_vars': False,
 'config': {},
 'displaycon': True,
 'displaylimit': 10,
 'dsn_filename': 'odbc.ini',
 'feedback': True,
 'parent': None,
 'polars_dataframe_kwargs': {},
 'short_errors': True,
 'style': 'DEFAULT'}

In [2]: self.trait_names()
Out[2]:
['autocommit',
 'autolimit',
 'autopandas',
 'autopolars',
 'column_local_vars',
 'config',
 'displaycon',
 'displaylimit',
 'dsn_filename',
 'feedback',
 'parent',
 'polars_dataframe_kwargs',
 'short_errors',
 'style']

In [3]: self.__dict__
Out[3]:
{'_trait_values': {'autocommit': True,
  'autolimit': 0,
  'autopandas': False,
  'autopolars': False,
  'column_local_vars': False,
  'displaycon': True,
  'displaylimit': 10,
  'dsn_filename': 'odbc.ini',
  'feedback': True,
  'parent': None,
  'short_errors': True,
  'style': 'DEFAULT'},
 '_trait_notifiers': {'config': {'change': [<traitlets.traitlets.ObserveHandler at 0x101582770>]},
  'autopandas': {'change': [<traitlets.traitlets.ObserveHandler at 0x12d9e0820>]},
  'autopolars': {'change': [<traitlets.traitlets.ObserveHandler at 0x12d9e0820>]}},
 '_trait_validators': {'displaylimit': <traitlets.traitlets.ValidateHandler at 0x12d9e0670>},
 '_cross_validation_lock': False,
 '_store': <sql.store.SQLStore at 0x103fe6b00>}

```

Reply:
Thanks for these comments. By using `trait_names()`, I was able to get the list of traits like @edublancas did in his comment. However, this includes `parent` and `config` that aren't `SqlMagic` configuration options in my understanding. So I still think removing those two is necessary. Please correct me if I'm not.

For later when you check the PR thoroughly, loading the `pyproject.toml` file isn't in the `SqlMagic.__init__`. It's inside the `load_ipython_extension`. I will move it to the `SqlMagic.__init__` before next review.

Reply:
ok, I looked at the code. I think it makes more sense to put if in load_ipython extension. but I saw that you're modifying the config by calling `ip.run_cell`, it'd be better to do it like this (you can add this in the `load_ipython_extension` function:

```python
In [36]: sql = ip.find_cell_magic("sql").__self__

In [39]: sql.autopandas
Out[39]: True

In [40]: sql.autopandas = False

In [41]: sql.autopandas
Out[41]: False
```

I think this will work but you'll have to check.

then you can implement the logic to read from the pyproject.toml and modify like this:

```python
setattr(sql, "autopandas", True)
```

Reply:
Made the change
- Using `trait_defaults` to get configuration settings is in `get_default_configs` inside `util.py`
- Using the following worked, and the change is in `set_configs` in `magic.py`.


> ```python
> In [36]: sql = ip.find_cell_magic("sql").__self__
> 
> In [39]: sql.autopandas
> Out[39]: True
> 
> In [40]: sql.autopandas = False
> 
> In [41]: sql.autopandas
> Out[41]: False
> ```